src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabMergeRequestCommentTrigger.java

```diff
@@ -0,0 +1,90 @@
+package io.jenkins.plugins.gitlabbranchsource;
+
+import hudson.model.CauseAction;
+import hudson.model.Job;
+import hudson.security.ACL;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import jenkins.model.ParameterizedJobMixIn;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.SCMSourceOwner;
+import jenkins.scm.api.SCMSourceOwners;
+import org.gitlab4j.api.webhook.NoteEvent;
+
+public class GitLabMergeRequestCommentTrigger extends AbstractGitLabJobTrigger<NoteEvent> {
+
+    public static final Logger LOGGER = Logger.getLogger(GitLabMergeRequestCommentTrigger.class.getName());
+
+    public GitLabMergeRequestCommentTrigger(NoteEvent payload) {
+        super(payload);
+    }
+
+    @Override
+    public void isMatch() {
+        if(getPayload().getObjectAttributes().getNoteableType().equals(NoteEvent.NoteableType.MERGE_REQUEST)) {
+            Integer mergeRequestId = getPayload().getMergeRequest().getIid();
+            final Pattern mergeRequestJobNamePattern = Pattern.compile("^MR-" + mergeRequestId + "\\b.*$",
+                    Pattern.CASE_INSENSITIVE);
+            final String commentBody = getPayload().getObjectAttributes().getNote();
+            final String commentUrl = getPayload().getObjectAttributes().getUrl();
+            ACL.impersonate(ACL.SYSTEM, () -> {
+                boolean jobFound = false;
+                for (final SCMSourceOwner owner : SCMSourceOwners.all()) {
```

Comment:
I suggest you try a lambda that filter `SCMSourceOwners` and map them to `SCMsources` and then filter `SCMSources`

Reply:
```java
@Override
    public void isMatch() {
        if(getPayload().getObjectAttributes().getNoteableType().equals(NoteEvent.NoteableType.MERGE_REQUEST)) {
            Integer mergeRequestId = getPayload().getMergeRequest().getIid();
            final Pattern mergeRequestJobNamePattern = Pattern.compile("^MR-" + mergeRequestId + "\\b.*$",
                    Pattern.CASE_INSENSITIVE);
            final String commentBody = getPayload().getObjectAttributes().getNote();
            final String commentUrl = getPayload().getObjectAttributes().getUrl();
            ACL.impersonate(ACL.SYSTEM, () -> {
                boolean jobFound = false;
                StreamSupport.stream(SCMSourceOwners.all().spliterator(), false)
                        .filter(owner -> !owner.getFullDisplayName().contains(" » "))
                        .map(SCMSourceOwner::getSCMSources)
                        .filter(scmSource -> scmSource instanceof GitLabSCMSource)
                        .forEach(scmSource -> {
                        GitLabSCMSource gitLabSCMSource = (GitLabSCMSource) scmSource;
                        final GitLabSCMSourceContext sourceContext = new GitLabSCMSourceContext(null, SCMHeadObserver.none())
                                .withTraits(gitLabSCMSource.getTraits());
                        if (!sourceContext
                                .mrCommentTriggerEnabled()) {
                            return;
                        }
                        if (gitLabSCMSource.getProjectId() == getPayload().getMergeRequest().getTargetProjectId()) {
                            for (Job<?, ?> job : owner.getAllJobs()) {
                                if (mergeRequestJobNamePattern.matcher(job.getName()).matches()) {
                                    String expectedCommentBody = sourceContext.getCommentBody();
                                        Pattern pattern = Pattern.compile(expectedCommentBody,
                                                Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
                                        if (commentBody == null || pattern.matcher(commentBody).matches()) {
                                            ParameterizedJobMixIn.scheduleBuild2(job, 0,
                                                    new CauseAction(new GitLabMergeRequestCommentCause(commentUrl)));
                                            LOGGER.log(Level.INFO,
                                                    "Triggered build for {0} due to MR comment on {1}",
                                                    new Object[] {
                                                            job.getFullName(),
                                                            getPayload().getProject().getPathWithNamespace()
                                                    }
                                            );
                                        } else {
                                            LOGGER.log(Level.INFO,
                                                    "MR comment does not match the trigger build string ({0}) for {1}",
                                                    new Object[] { expectedCommentBody, job.getFullName() }
                                            );
                                        }
                                        break;
                                    }
                                    jobFound = true;
                                }
                            }
                        });
                if (!jobFound) {
                    LOGGER.log(Level.INFO, "MR comment on {0} did not match any job",
                            new Object[] {
                                    getPayload().getProject().getPathWithNamespace()
                            }
                    );
                }
            });
        }
    }
```

There are two issues with this impl:

1. `jobFound` variable should be final or effectively final
2. Cannot access `owner` inside `forEach` of SCMSources.